### PR TITLE
[FrameworkBundle] changed configuration file for messenger from xml to php

### PR DESCRIPTION
**Pull request by hyoa**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       |  https://github.com/symfony/symfony/issues/37186
| License       | MIT

Use `messenger.php` instead of `messenger.xml` and `messenger_debug.php` instead of `messenger_debug.php`

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 1]
```diff
@@ -1629,13 +1629,13 @@ private function registerLockConfiguration(array $config, ContainerBuilder $cont
         }
     }
 
-    private function registerMessengerConfiguration(array $config, ContainerBuilder $container, XmlFileLoader $loader, array $validationConfig)
+    private function registerMessengerConfiguration(array $config, ContainerBuilder $container, XmlFileLoader $loader, array $validationConfig, PhpFileLoader $phpLoader)
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:1632]: `XmlFileLoader` is not used in this function anymore. please just change the typhint from XmlFileLoader to PhpFileLoader  without a new argument.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger_debug.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+namespace Symfony\Component\DependencyInjection\Loader\Configurator;
+
+use Symfony\Component\Messenger\DataCollector\MessengerDataCollector;
+
+return static function (ContainerConfigurator $container) {
+    $container->services()
+        ->set('data_collector.messenger', MessengerDataCollector::class)
+            ->tag(
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger_debug.php:19]: please use this coding style:
```
->tag('data_collector', [
    ...
])

**Review by Tobion — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        ->set('messenger.retry_strategy_locator')
+            ->args([
+                [],
+            ])
+            ->tag('container.service_locator')
+
+        ->set('messenger.retry.abstract_multiplier_retry_strategy', MultiplierRetryStrategy::class)
+            ->abstract()
+            ->args([
+                '', // max retries
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php:135]: all empty arguments with comments should use `abstract_arg`, e.g.  `abstract_arg('max retries')`

**Review by Tobion — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        ->set('messenger.middleware.traceable', TraceableMiddleware::class)
+            ->abstract()
+            ->args([
+                service('debug.stopwatch'),
+            ])
+
+        // Discovery
+        ->set('messenger.receiver_locator')
+            ->args([
+                [],
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php:99]: abstract_arg pls

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        ->set('messenger.transport.in_memory.factory', InMemoryTransportFactory::class)
+            ->tag('messenger.transport_factory')
+            ->tag('kernel.reset', ['method' => 'reset'])
+
+        ->set('messenger.transport.sqs.factory', AmazonSqsTransportFactory::class)
+
+        // retry
+        ->set('messenger.retry_strategy_locator')
+            ->args([
+                [],
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php:128]: abstract_arg pls

**Review by Tobion — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php (hunk 5 above)
Comment (hyoa) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php:128]: What sould I put in the method argument as there is no comment ?

Comment (Tobion) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php:128]: retry strategy services

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php (hunk 4 above)
Comment (Tobion) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php:99]: you can put "receiver name to service mapping"

Comment (hyoa) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php]: test fail with this changement

```
Invalid definition for service "messenger.receiver_locator": an array of references is expected as first argument when the "container.service_locator" tag is set.
```

Comment (Tobion) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php]: The problem is probably that the compiler pass to handle container.service_locator tags, validates the argument. But the service is not used at all and would be removed. So the validation should skip abstract_arg. I wonder if the compiler pass does not run after unused services are removed?

Anyway, please don't use abstract_arg then for messenger.receiver_locator and messenger.retry_strategy_locator. Fixing this should be separate thing.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php (hunk 5 above)
Comment (Tobion) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php]: please revert due to failing tests

**Review by Tobion — CHANGES_REQUESTED** [review]
please don't use abstract_arg then for messenger.receiver_locator and messenger.retry_strategy_locator due to the failing tests

**Comment by hyoa** [comment]
I took a look at the travis error.
From what I understood, 

* when `messenger.default_serializer` was created using `messenger.xml`, it look like:

```php
Symfony\Component\DependencyInjection\Alias^ {#3889
  -id: "messenger.default_serializer"
  -public: false
  -private: false
  -deprecation: []
}
```

* when created with `messenger.php` it look like

```php
  "Symfony\Component\Messenger\Transport\Serialization\SerializerInterface" => Symfony\Component\DependencyInjection\Alias^ {#3006
    -id: "messenger.default_serializer"
    -public: true
    -private: true
    -deprecation: []
  }
```

Changing `->alias(SerializerInterface::class, 'messenger.default_serializer')` by `->alias(SerializerInterface::class, 'messenger.default_serializer')->private()` fix the test, but I'm not sure if it is the good thing to do

**Comment by nicolas-grekas** [comment]
I pushed a fix to make tests green with this patch:
image

**Review by Tobion — APPROVED** [review]

**Comment by Tobion** [comment]
Thank you @hyoa.

**Pull request merged**